### ISSUE TITLE
multiply by a transpose type

### ISSUE THREAD

**Issue by JeffFessler** [Jutho/LinearMaps.jl#147]
The following code fails in LM v0.3.3 with Julia v1.6.1:

```julia
using LinearAlgebra 
using LinearMaps

M = ones(3,3)
A = LinearMap(M)
S = LinearAlgebra.Symmetric(M)
T = LinearAlgebra.Transpose(S)
@assert T isa AbstractMatrix # true
M * T 
A * T # fails
```

The type of `T` is a strange one:
`3×3 transpose(::Symmetric{Float64, Matrix{Float64}}) with eltype Float64`
and frankly I don't even know why Julia has this type.
But the type is there in LinearAlgebra and it is an `AbstractMatrix` so `A * T` should work!

The error message involves `cholesky` which is baffling given that this is just matrix multiplication,
so maybe the error is Base, but `M * T` works fine so I'm starting here.

**Comment by dkarrasch** [comment]
Interesting case. I think the root cause is your "unorthodox" usage of `LinearAlgebra.Transpose`. This is a constructor of the `Transpose` wrapper type. You should use it only in library code, to wrap your custom matrix types and things like that. If you want to just transpose your matrix, you should use `transpose(S)`. That function decides via multiple dispatch whether it should wrap the argument in a `Transpose` or do something else. For instance, if you call `transpose(D::Diagonal)`, it will return a `Diagonal` with `transpose`d elements (due to the recursive nature). Similarly, `transpose(S::Symmetric)` is a noop and will simply return `S`. In that case, you'll get
```julia
julia> A*T
3×3 LinearMaps.CompositeMap{Float64} with 2 maps:
  3×3 LinearMaps.WrappedMap{Float64} of
    3×3 Symmetric{Float64, Matrix{Float64}}
  3×3 LinearMaps.WrappedMap{Float64} of
    3×3 Matrix{Float64}
```
hopefully as desired.

The error you're getting has to do with the following: you multiply a `LinearMap` by a matrix. In that case, the matrix is supposed to get wrapped by a `LinearMap`, whose symmetry/definiteness properties need to be determined (by default). And the `isposdef` check is where the `cholesky` comes in and fails for your exotic type construction. 😉

**Comment by JeffFessler** [comment]
I really appreciate you taking the time to make that detailed reply.

I kind of guessed that you would say that.  So now let me explain the root cause of this.
I am trying to eliminate method ambiguities over here:
https://github.com/JeffFessler/LinearMapsAA.jl/pull/25
Annoyingly, this `Transpose` type was causing a method ambiguity.
(I had never heard of or used `Transpose` prior to this.)
The type may be "exotic" but still LinearAlgebra has a method in it for multiplying such a type by another AbstractMatrix so it is a type they expect to be used in multiply operations, and hence fair game for LM?

I resolved the ambiguity by adding a specialized method for it, but then to get full code coverage I need to test that method, but sadly the test fails because of the issue in the MWE above.
(I understand that normally one would just use `transpose` but using `transpose` does not generate the appropriate type to cover the code in question.)

I was hoping you would want LMs to be able to multiply by _any_ type of `AbstractMatrix`.
But now it seems to me that the issue is really with Base, not with LM, because if this type is there then there really should be a way to apply `cholesky` to it.  The issue seems to be that the type annotation here is too restrictive:
https://github.com/JuliaLang/julia/blob/6aaedecc447e3d8226d5027fb13d0c3cbfbfea2a/stdlib/LinearAlgebra/src/cholesky.jl#L377

Meanwhile, your helpful reply gives me a tolerable workaround for now!
I am simply adding this to my test routine:
`LinearAlgebra.isposdef(A::Transpose{Float64, Symmetric{Float64, Matrix{Float64}}}) = false`

BTW, I had never realized that making a `CompositeMap` might involve an expensive call like `isposdef`.
Is that really the default you want?  I think such maps as being a "lazy" product (in the good sense) and calling `cholesky` seems the opposite of lazy.  And it seems like undesirable overhead because even if A and B are both positive definite, it is rare that the product A * B is, so wouldn't it be OK to just make the default that posdef=false whenever wrapping something to go into a CompositeMap?

**Comment by dkarrasch** [comment]
Wait, I didn't mean to say that the `Transpose` type is exotic. It is a very useful type, of course. I meant that the `Transpose{Symmetric}` is exotic, and kind of non-sensical, because symmetric matrices are invariant under transposition, so why would one add a `Transpose` annotation/wrapper to an inherently `Symmetric` object? 

> (I understand that normally one would just use transpose but using transpose does not generate the appropriate type to cover the code in question.)

I need to understand better where this type is created. I think it shouldn't even be created in the first place.

> I was hoping you would want LMs to be able to multiply by any type of AbstractMatrix.

Well, that's what we have defined in principle. It fails for an "unrelated" reason: you could simply call `LinearMap(A::Transpose{...,Symmetric{...}})` and that would fail for the exact same reason. It's the constructor that is "failing", and multiplication fails because it's calling the constructor without the `isposdef` property set manually. If that's generally a good idea, that may be debatable.

> I am simply adding this to my test routine:
`LinearAlgebra.isposdef(A::Transpose{Float64, Symmetric{Float64, Matrix{Float64}}}) = false`

I'd make it exact:
```julia
LinearAlgebra.isposdef(A::Transpose{Float64, Symmetric{Float64, Matrix{Float64}}}) = isposdef(parent(A))
```

if you really need to.

> BTW, I had never realized that making a `CompositeMap` might involve an expensive call like `isposdef`.

You may have a good point here, I don't know. So far, nobody complained. 😄 "To our defense", that check is very cheap for non-square matrices. Also, that check can be avoided (in library code, where you have control over the circumstances) by wrapping the matrix manually. So, instead of `L::LinearMap * A::AbstractMatrix` you call `L * LinearMap(A; isposdef=true/false)`.

>  The issue seems to be that the type annotation here is too restrictive:

I'm not sure. That `Transpose{Symmetric}` type is not expected to occur...

> so wouldn't it be OK to just make the default that posdef=false whenever wrapping something to go into a CompositeMap?

That seems to make sense! It's a different story for automatic promotion in `+`, but for `*` that seems to be a more reasonable default. I'll play with that.

**Comment by JeffFessler** [comment]
> I need to understand better where this type is created. I think it shouldn't even be created in the first place.

I tend to agree. Here are the lines of code that caused my method ambiguity:
https://github.com/JuliaLang/julia/blob/752ac81c8f8f6023a6a01a629ebcfa5de75b7266/stdlib/LinearAlgebra/src/symmetric.jl#L607

```julia
*(A::AbstractMatrix, transB::Transpose{<:Any,<:RealHermSymComplexSym}) = A * transB.parent
*(transA::Transpose{<:Any,<:RealHermSymComplexSym}, B::AbstractMatrix) = transA.parent * B
*(adjA::Adjoint{<:Any,<:RealHermSymComplexHerm}, B::AbstractMatrix) = adjA.parent * B
*(A::AbstractMatrix, adjB::Adjoint{<:Any,<:RealHermSymComplexHerm}) = A * adjB.parent
```
You might recognize one of the co-authors of that file of code  😄

I don't fully understand this `RealHermSymComplexSym` type and its `Transpose` or `Adjoint`, but obviously someone thought there should be a multiplication method for it!
```julia
julia> LinearAlgebra.RealHermSymComplexSym
Union{Hermitian{T, S}, Symmetric{T, S}, Symmetric{Complex{T}, S}} where {T<:Real, S}
```
If those methods were not there, I would not have opened this issue because I never would have used `Transpose` directly.
I wonder if there is code coverage for those methods in LinAlg?
Possibly not given the "exotic" lengths one must go to create such an object.




> I'd make it exact:
> 
> ```julia
> LinearAlgebra.isposdef(A::Transpose{Float64, Symmetric{Float64, Matrix{Float64}}}) = isposdef(parent(A))
> ```

Oh that's nice!  Should it be in a PR for LA?  Even more generally:
```julia
LinearAlgebra.isposdef(A::Transpose) = isposdef(parent(A))
```

**Comment by dkarrasch** [comment]
As for the automatic `isposdef` check upon multiplying a `LinearMap` with an `AbstractMatrix`: in your use cases, are your matrices square? If not, then that check quits after the size checks. With your proposal of strengthening the `isposdef` check for `CompositeMap` it may be really helpful to know if the matrix is posdef. One way to reduce the amount of implicit work is to make
```julia
*(A::LinearMap, B::AbstractMatrix) = A * WrappedMap(B, isposdef=false)
```
The call to `WrappedMap` would then still check for symmetry, but that's usually quick, especially for special matrix types. The downside is that this will then fail to recognize `B * A * B'` (for posdef `B::AbstractMatrix`)  as positive definite, rendering your strengthened sufficient criterion somewhat (almost) useless (unless you happen to work with `LinearMap`s only, whose properties have been carefully determined/set.

**Comment by JeffFessler** [comment]
I do have square use cases, like the product of a DFT matrix with a Diagonal matrix (where the Diagonal gets wrapped).  (This arises in phase retrieval with a coded-aperture mask.)  But honestly I never use `isposdef` and it would be OK with me to eliminate the probably-nearly-useless sufficient condition in #151 and simply default to `false`.  Mathematically it kind of bothers me to call it `false` - in theory I'd prefer an enum type of `true` or `false` or, say, `missing` to represent the (common!) case where the code can't determine easily, but that would be too much work in practice.  Hopefully users understand that `false` == unknown here.

**Comment by dkarrasch** [comment]
>  Hopefully users understand that `false` == unknown here.

I'm sure they do. Our checks have always been of sufficient kind. We may have communicated that poorly in the docs, but it's been there (as a comment) in the code ever since.

**Comment by dkarrasch** [comment]
I forgot what we have discussed here and what the cause of the issue was, but the code in the OP no longer fails, so I'm closing this. Please reopen if something has not yet been covered.

**Comment by JeffFessler** [comment]
Yes, I think this was addressed by #151 and/or #154.
And (mainly as a reminder to me) perhaps also related to
https://github.com/JuliaLang/julia/pull/41188
Thanks!
